src/Raven.Server/Utils/MetricCacher.cs

```diff
@@ -67,86 +68,55 @@ public T GetValue<T>(string key, Func<T> factory = null)
             if (_metrics.TryGetValue(key, out var value) == false)
                 throw new InvalidOperationException($"Metric '{key}' was not found.");
 
-            if (value.ShouldRefresh(out var first) == false)
-                return (T)value.Value;
-
-            if (first)
-            {
-                lock (value)
-                {
-                    if (value.ShouldRefresh(out first))
-                        value.Refresh();
-
-                    return (T)value.Value;
-                }
-            }
-
-            if (Monitor.TryEnter(value, 0) == false)
-                return (T)value.Value;
-
-            try
-            {
-                if (value.ShouldRefresh(out first))
-                    value.Refresh();
-
-                return (T)value.Value;
-            }
-            finally
-            {
-                Monitor.Exit(value);
-            }
+            return (T)value.GetRefreshedValue();
         }
 
-        private class MetricValue<T> : MetricValueBase
+        private class MetricValue
         {
-            private readonly Func<T> _factory;
+            private readonly TimeSpan _refreshRate;
+            private Func<object> _factory;
+            private Task<DateTime> _task;
+            private object _value;
 
-            public MetricValue(TimeSpan refreshRate, Func<T> factory)
-                : base(refreshRate)
+            public MetricValue(TimeSpan refreshRate, Func<object> factory)
             {
-                if (factory == null)
-                    throw new ArgumentNullException(nameof(factory));
-
+                _refreshRate = refreshRate;
                 _factory = factory;
+                _value = factory();
+                _task = Task.FromResult(SystemTime.UtcNow + _refreshRate);
             }
 
-            protected override object RefreshInternal()
-            {
-                return _factory();
-            }
-        }
 
-        private abstract class MetricValueBase
-        {
-            private DateTime _lastRefresh;
-            private TimeSpan _refreshRate;
-
-            protected MetricValueBase(TimeSpan refreshRate)
+            public object GetRefreshedValue()
             {
-                _refreshRate = refreshRate;
-            }
+                var currentTask = _task;
+                if (currentTask.IsCompleted == false)
+                    return _value; // return current value, while it is being computed
 
-            public object Value { get; private set; }
+                if (currentTask.IsFaulted)
```

Comment:
this is problematic because inability to start a task (e.g. low memory) or any error, will cause infinite exceptions here, so basically you will force users to restart the server

Reply:
You are correct, fixed.